File: 01-write-your-own-simple-http-server/golang/main.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"strings"
)

func main() {
	ln, err := net.Listen("tcp", ":8080")
	if err != nil {
		panic(err)
	}
	fmt.Println("listening on" + ln.Addr().String())

	defer ln.Close()

	for {
		conn, err := ln.Accept()
		if err != nil {
			fmt.Println(err)
			return
		}

		go func() {
			fmt.Println("connected to " + conn.RemoteAddr().String())

			defer conn.Close()

			sc := bufio.NewScanner(conn)
			for !sc.Scan() {
			}

			s := strings.Split(sc.Text(), " ")
			fmt.Printf("%#v", s)
			if len(s) < 3{
				fmt.Println("invalid request")
				return
			}

			method, path := s[0], s[1]

			if method == "GET" {
				fmt.Fprint(conn, "HTTP/1.1 200 OK\r\n")
				fmt.Fprint(conn, "Content-Type: text/html\r\n")
				fmt.Fprint(conn, "\r\n")
				fmt.Fprint(conn, "<p>anda mengakses ", path, "</p>")

			} else {
				fmt.Fprint(conn, "HTTP/1.1 405 Method Not Allowed\r\n")
				fmt.Fprint(conn, "\r\n")
			}
		}()
	}
}

File: 01-write-your-own-simple-http-server/rust/src/main.rs
use tokio::io::{AsyncWriteExt, BufReader};
use tokio::{io::AsyncBufReadExt, net::TcpListener};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("Hello, world!");
    let ln = TcpListener::bind("localhost:8080").await?;
    println!("Listening on {}", ln.local_addr()?);
    loop {
        let (stream, addr) = ln.accept().await?;
        let mut stream = BufReader::new(stream);

        println!("connected to {}", addr);
        tokio::spawn(async move {
            let mut line = String::new();
            if let Err(..) = stream.read_line(&mut line).await {
                println!("error reading line");
                return;
            }
            let s: Vec<&str> = line.split_whitespace().collect();
            if s.len() < 3 {
                println!("invalid request");
                return;
            }
            let (method, path) = (s[0], s[1]);
            if method == "GET" {
                if let Err(..) = stream
                    .write_all(
                        format!(
                            "<HTTP/1.1 200 OK\r\n\
                    Content-Type: text/html\r\n\
                    \r\n\
                    <p>anda mengakses {}</p>",
                            path
                        )
                        .as_bytes(),
                    )
                    .await
                {
                    println!("error writing response");
                    return;
                }
            } else {
                if let Err(..) = stream
                    .write_all(
                        b"<HTTP/1.1 405 Method Not Allowed\r\n\
                    \r\n",
                    )
                    .await
                {
                    println!("error writing response");
                    return;
                }
            }
        });
    }
}
